### PR TITLE
[Codegen] Add ops to translate between tensor and memref

### PR DESCRIPTION
**Pull request by Max191**
This PR adds 2 new ops to go from tensors to memrefs and memrefs to tensors. The new ops are:

 - `iree_codegen.load_from_memref` - Loads a tensor from a memref with the same shape and element type.
 - `iree_codegen.store_to_memref` - Stores a tensor to a memref with the same shape and element type.

These ops can be used to bufferize hal.interface.subspan and iree_tensor_ext.dispatch.tensor.store ops early on, which allows for certain transformations that are only possible with memref semantics to happen without needing to bufferize the entire dispatch.

This PR only adds the basic op definitions, and various op interface implementations (bufferization, subset insertion, etc.) will come as follow-ups.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenOps.td [hunk 1]
```diff
@@ -171,5 +171,54 @@ def IREECodegen_NullPointerOp :
...
+  let description = [{
+    Loads a tensor from a memref with the same shape and element type. The
+    read_only attribute indicates that the source buffer from which the tensor
+    is read is a read only buffer. This hint can be used by bufferization to
+    determine whether or not the buffer that this op reads from may be written.
+  }];
+
+  let arguments = (ins
+    AnyStridedMemRef:$source,
+    UnitAttr:$read_only
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenOps.td:191]: IIUC, we don't set the flag if and only if it is a readwrite buffer? Instead of carrying the information in the op, can we just look at the `source`? Is it because we're losing the information when we fold those indexing/slicing/reshaping ops into the memrefs?

File: compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/test/roundtrip.mlir [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+}
+// CHECK-LABEL: func.func @load_from_memref(
+// CHECK-SAME:    %[[ARG0:[a-zA-Z0-9_]+]]
+// CHECK:         iree_codegen.load_from_memref %[[ARG0]]
+// CHECK-SAME:      : memref<4xf32> -> tensor<4xf32>
+
+// -----
+
+func.func @load_from_memref_mixed_static_dynamic(%arg0: memref<?x4xf32>) -> tensor<4x?xf32> {
+  %value = iree_codegen.load_from_memref %arg0 : memref<?x4xf32> -> tensor<4x?xf32>
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/test/roundtrip.mlir:26]: It looks weird to me that we allow compatible shapes, but not require them being identical shape, because the load op does not take any other operands like indices. The intention of the op is more like loading the whole value from the memref type as a tensor value. It is a specialized load to me. Is there a reason to support compatible shapes?

File: compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/test/roundtrip.mlir
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/test/roundtrip.mlir:1]: We are missing a roundtrip test for the load op with readonly attribute.

File: compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenOps.td [hunk 3]
```diff
@@ -171,5 +171,54 @@ def IREECodegen_NullPointerOp :
...
+  }];
+
+  let hasVerifier = 1;
+}
+
+def IREECodegen_StoreToMemrefOp : Op<IREECodegen_Dialect, "store_to_memref",
+    [DeclareOpInterfaceMethods<MemoryEffectsOpInterface>]> {
+  let summary = [{stores a tensor into a memref}];
+  let description = [{
+    Stores a tensor into a memref with the same shape and element type.
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenOps.td]: maybe replace "same" with "compatible" if we are going to allow compatible shapes.

```suggestion
    Stores a tensor into a memref with the compatible shape and the same element type.
```

**Review by hanhanW — COMMENTED** [review]
What is subset insertion? Do we have such interface?

File: compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenOps.td (hunk 1 above)
Comment (Max191) [reply compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenOps.td:191]: Yes, it will be set for readonly, and not set for readwrite. The flag is just a hint for bufferization of the op, which does not look at the memref operands. I think it makes sense to keep it on the op, since it is about the tensor result, and it is just used for the bufferization of the operation.

If we tried to infer from the subspan that produces the source, then it could be tricky because we can have other ops in between. Also, the op is not meant to have anything to do with the subspan, since it is just a translation from any memref to a tensor.

File: compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/test/roundtrip.mlir (hunk 2 above)
Comment (Max191) [reply compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/test/roundtrip.mlir:26]: Identical shapes seems fine to me too. I made it compatible shapes because of some strange lit tests with dispatch.tensor.load ops that produced unnecessary dynamic dimensions, but I think we can still support those cases using subviews. Using compatible shapes for this op was mostly about allowing for these weird corner cases, but perhaps it is less error prone with equal shapes.

**Comment by Max191** [comment]
> What is subset insertion? Do we have such interface?

Yes, it is an interface upstream: https://github.com/llvm/llvm-project/blob/10ea5eedb1d13760e4caf0bd983ce7778cff0426/mlir/include/mlir/Interfaces/SubsetOpInterface.td#L150

It is mainly used in empty tensor elimination to get more in place bufferization: https://github.com/llvm/llvm-project/blob/10ea5eedb1d13760e4caf0bd983ce7778cff0426/mlir/include/mlir/Dialect/Bufferization/Transforms/Transforms.h#L24-L58

File: compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenOps.cpp [hunk 4]
```diff
@@ -65,3 +65,45 @@ void ExtractStridedMetadataOp::getAsmResultNames(
...
 }
+
+//===----------------------------------------------------------------------===//
+// LoadFromMemrefOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult LoadFromMemrefOp::verify() {
+  RankedTensorType resultType = getResult().getType();
+  MemRefType sourceType = getSource().getType();
+  if (!llvm::equal(resultType.getShape(), sourceType.getShape()) ||
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenOps.cpp]: nit: I think `==` works just fine on arrayrefs

File: compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenOps.cpp [hunk 5]
```diff
@@ -65,3 +65,45 @@ void ExtractStridedMetadataOp::getAsmResultNames(
...
+}
+
+//===----------------------------------------------------------------------===//
+// StoreToMemrefOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult StoreToMemrefOp::verify() {
+  RankedTensorType valueType = getValue().getType();
+  MemRefType targetType = getTarget().getType();
+  if (!llvm::equal(valueType.getShape(), targetType.getShape()) ||
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenOps.cpp]: also here

File: compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/test/roundtrip.mlir (hunk 2 above)
Comment (Max191) [reply compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/test/roundtrip.mlir:26]: Actually, after making this change, I am running into issues with the corner case because the subview op verifier does not allow dynamic dims when the sizes are static. This is the problem case:

```
%3 = iree_tensor_ext.dispatch.tensor.load %1, offsets = [0], sizes = [4], strides = [1]
      : !iree_tensor_ext.dispatch.tensor<readonly:tensor<?xf32>>{%binding_size} -> tensor<?xf32>
```
This op passes the verifier, but it is not possible to make a subview that with a size of `4` that produces a dynamic shape memref, because the verifier will fail.

I think we can either allow compatible shapes in iree_codegen.load_from_memref/store_to_memref, or we can update the verifier of dispatch.tensor.load/store ops. I doubt we ever generate this IR, so it is probably fine to change the dispatch.tensor.load/store verifiers. WDYT?

Comment (Max191) [reply compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/test/roundtrip.mlir:26]: The DispatchTensorLoadOp has a very minimal verifier:
https://github.com/iree-org/iree/blob/8069d7b1f47c73d997c54c8a179cbd60f4ca23da/compiler/src/iree/compiler/Dialect/TensorExt/IR/TensorExtOps.cpp#L128-L134

And the DispatchTensorStoreOp is doing some slice verification:
https://github.com/iree-org/iree/blob/8069d7b1f47c73d997c54c8a179cbd60f4ca23da/compiler/src/iree/compiler/Dialect/TensorExt/IR/TensorExtOps.cpp#L300-L314

I think it makes the most sense to update the DispatchTensorLoadOp verifier to be more consistent with DispatchTensorStore, so I think we can leave these new ops as requiring shapes to be equal, and not compatible.

Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/test/roundtrip.mlir:26]: Allowing compatible shape may be intended because we have [DispatchTensorLoadOpWithOffsetSizesAndStridesConstantArgumentFolder](https://github.com/iree-org/iree/blob/b4e3694a31a92a814a7b127f288e5928705fb3c4/compiler/src/iree/compiler/Dialect/TensorExt/IR/TensorExtOpFolders.cpp#L139-L164) pattern, although I think we should restrict it like other load/store operations. We could run into rabbit hole if we make the change. Let's go with compatible shape for now, and we can always infer the shape when needed. Thanks for taking a deep look at this!

Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/test/roundtrip.mlir:26]: I just recall that I asked Ben a similar question before when I added the flow/stream encode ops. I think we can allow it if we can generate code for it.

EDIT: the question was: do we allow encode `tensor<?x4xf32>` to `tensor<5x?xf32, #encoding>`? 

https://discord.com/channels/689900678990135345/1254843174111678555/1350206891090907177

Comment (Max191) [reply compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/test/roundtrip.mlir:26]: I see, I'll make it use compatible shapes then.

**Review by hanhanW — APPROVED** [review]
